src/execution/BucketExecutor.tsx

```diff
@@ -0,0 +1,338 @@
+// (C) 2007-2018 GoodData Corporation
+// tslint:disable:member-ordering
+import * as React from "react";
+import { SDK, factory as createSdk } from "@gooddata/gooddata-js";
+import { AFM, Execution, VisualizationObject, VisualizationInput } from "@gooddata/typings";
+
+import { setTelemetryHeaders } from "../helpers/utils";
+import { convertErrors } from "../helpers/errorHandlers";
+import { RuntimeError } from "../errors/RuntimeError";
+import { convertBucketsToAFM } from "../helpers/conversion";
+import { MEASURES, ATTRIBUTES } from "../constants/bucketNames";
+import { MEASUREGROUP } from "../internal/constants/bucket";
+
+type dimensionItem = VisualizationInput.AttributeOrMeasure | VisualizationInput.IAttribute;
+
+export const getExecutionFromDimensions = (
+    dimensions: dimensionItem[][],
+    filters: VisualizationInput.IFilter[],
+    sorts: VisualizationInput.ISort[],
+): AFM.IExecution => {
+    const measuresBucket: VisualizationObject.IBucket = {
+        localIdentifier: MEASURES,
+        items: [],
+    };
+    const attributesBucket: VisualizationObject.IBucket = {
+        localIdentifier: ATTRIBUTES,
+        items: [],
+    };
+
+    const resultSpec: AFM.IResultSpec = {
+        dimensions: [],
+        sorts,
+    };
+
+    let measureDimension: number | null = null;
+
+    dimensions.forEach((dimension, dimensionIndex) => {
```

Comment:
This whole block could be extracted to a function -> could be tested independently of react.